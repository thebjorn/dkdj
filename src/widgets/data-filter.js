

/*
 *   This is the top-level filter widget, which groups all sub-filters
 *   for a given data table/grid.
 */
// const _ = require('lodash');
// const dk = require('../boot');
// const Widget = require('../widgetcore/dk-widget.js');
// const forms = require('../forms');

import dk from "../dk-obj";
import {Widget} from "../widgetcore/dk-widget";
import {Template} from "../browser/dk-dom-template";
import {CheckboxSelectWidget, RadioSelectWidget, TriboolWidget} from "../forms/widgets";
import {zip_object} from "../collections";


export function list2data (lst) {
    if (Array.isArray(lst)) {
        const res = {};
        lst.forEach(function (item) {
            res[item] = item;
        });
        res.__order = lst;      // XXX: Not needed anymore
        return res;
    } else {
        return lst;
    }
}


/*
 *  The top level filter container/widget.
 *
 *  dk.filter.DataFilter.create_on(filterbox, {
 *      ...
 *      filters: {
 *          testleader: {
 *              label: 'Testleder',
 *              select_multiple: true,
 *              url: 'progress/all!get-testleader-values'
 *          },
 *          favyear: {
 *              label: 'Favoritt√•r',
 *              values: [2013, 2014, 2015, 2016],
 *              select_multiple: true
 *          },
 *          smile: {
 *              label: 'Smil',
 *              value: undefined,
 *              construct(location) {
 *                  const self = this;
 *                  const btn = $('<button/>').text(":-)");
 *                  this.widget = location.append(btn);
 *                  $(this.widget).on('click', function () {
 *                      self.value = "*SMILE*";
 *                      dk.publish(self.widget, 'change', self.widget);
 *                  });
 *              }
 *          }
 *      }
 *  });
 */
export class DataFilter extends Widget {
    constructor(...args) {
        super({
            // type: 'DataFilter',
            data: undefined,
            heading: 'Filter',
            filters: [],         // XXX shouldn't this be {} or null?
            dataset: null,
            structure: {
                header: {
                    title: {
                        text: 'Filter'
                    }
                },
                content: {}
            },

            template: {
                root: 'section',
                header: '<header/>',
                title: '<h2/>',
                content: '<div/>',

                // Each filter is contained in a .filterbox that is structured
                // similarly to the main filter section.
                filterbox: '<div/>',
                filterheader: '<header/>',
                filtertitle: '<h4/>',
                filtercontent: '<div class="filter content"/>'
            },
            
        }, ...args);
    }

    /*
     *  construct_filterbx is called for each filter definition in this.filters
     *  and creates the radio/check boxes for the filters, calls the data-set
     *  to fetch all filter values, etc.
     */
    construct_filterbx(location, filter, filter_name) {
        const self = this;
        const structure = {
            filterbox: {
                classes: [filter_name],
                filterheader: {
                    filtertitle: {
                        text: filter.label || filter_name
                    }
                },
                filtercontent: {}
            }
        };
        const templ = Template(structure, this.template);
        const dom = templ.append_to(location, this, location);

        if (!filter.construct) {
            // Filters that don't construct themselves are constructed either as single or
            // multiple selects. Default is single-selection (i.e. radio selects).
            let widgetclass = filter.select_multiple ? CheckboxSelectWidget : RadioSelectWidget;
            if (filter.tribool) {
                widgetclass = TriboolWidget;
            }

            filter.construct = function (loc) {
                // this here refers to the filter, which we are attaching a new method to..
                const me = this;

                // noinspection JSPotentiallyInvalidUsageOfClassThis
                this.widget = widgetclass.create_inside(loc, {
                    data: list2data(this.values),
                    name: filter_name,
                    label: me.label
                });

                if (!this.widget.data) {
                    self.dataset.get_filter_data(filter_name, this.widget.FN('draw'));
                }
            };
        }
        filter.construct(location.filterbox.filtercontent);
        if (filter.widget) {
            // filters with their own `construct()` don't necessarily have `.widget`
            dk.on(filter.widget, 'change').run(function () {
                dk.debug('filter-change', self.values());
                if (self.data) self.data.set_filter(self.values());
                dk.publish(self, 'filter-change', self.values());
            });
        }
        return dom;
    }

    construct() {
        this.filters.forEach((value, key) => {
            //console.log('key:', key, 'value:', value);
            this.construct_filterbx(this.content, value, key);
        });
    }

    values() {
        const self = this;
        const keys = Object.keys(this.filters);
        const vals = keys.map(function (filter_name) {
            const filter = self.filters[filter_name];
            if (filter.value) {
                // filter has defined a value on itself..
                return filter.value;
            }
            if (filter.widget && filter.widget.jquery) {
                // found a jQuery object, get its .val()
                return filter.widget.val();
            }
            if (filter.widget && filter.widget.value) {
                // found a dk.form object
                return filter.widget.value;
            }
        });
        // dk.dir(_.object(keys, vals));
        return zip_object(keys, vals);
    }
}
