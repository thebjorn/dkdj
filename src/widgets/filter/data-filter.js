/*
 *   This is the top-level filter widget, which groups all sub-filters
 *   for a given data table/grid.
 */

import dk from "../../dk-obj";
import {zip_object} from "../../collections";
import {UIWidget} from "../../widgetcore/ui-widget";
import is from "../../is";
import {dkconsole} from "../../lifecycle/dkboot/dk-console";
import {CustomFilterDef, SelectMultipleFilterDef, SelectOneFilterDef} from "./filter-defs";



/*
 *  The top level filter container/widget.
 *
 *  dk.filter.DataFilter.create_on(filterbox, {
 *      ...
 *      filters: {
 *          testleader: {
 *              label: 'Testleder',
 *              select_multiple: true,
 *              url: 'progress/all!get-testleader-values'
 *          },
 *          favyear: {
 *              label: 'Favoritt√•r',
 *              values: [2013, 2014, 2015, 2016],
 *              select_multiple: true
 *          },
 *          smile: {
 *              label: 'Smil',
 *              value: undefined,
 *              construct(location) {
 *                  const self = this;
 *                  const btn = $('<button/>').text(":-)");
 *                  this.widget = location.append(btn);
 *                  $(this.widget).on('click', function () {
 *                      self.value = "*SMILE*";
 *                      dk.publish(self.widget, 'change', self.widget);
 *                  });
 *              }
 *          }
 *      }
 *  });
 */
export class DataFilter extends UIWidget {
    constructor(...args) {
        super({
            // type: 'DataFilter',
            data: undefined,
            heading: 'Filter',
            filters: {},
            dataset: null,
            structure: {
                header: {
                    title: {
                        text: 'Filter'
                    }
                },
                content: {}
            },

            template: {
                root: 'section',
                header: '<header/>',
                title: '<h2/>',
                content: '<div/>',

                // Each filter is contained in a .filterbox that is structured
                // similarly to the main filter section.
                filterbox: '<div/>',
                filterheader: '<header/>',
                filtertitle: '<h3/>',
                filtercontent: '<div class="filter content"/>'
            },
            
        }, ...args);
        
        this._filter_data = this._get_filter_data(this.filters);
    }

    /**
     *  Convert the user supplied `.filters` to a list of FilterDefBase 
     *  subclasses, which can (later) be combined with a location to create
     *  a Widget.
     *  
     *  filterdata filterdef is either
     *    
     *    (a) { select_multiple: true, values: .., ... }
     *    (b) { construct(location) {}, ... }
     *    (c) { values: .., ... }
     *   
     *  For option (b), the object must 
     *  
     *    (i) create a `this.widget` property (the filter system listens
     *        for notifications on this widget.
     *   (ii) set `this.value` to the correct filter value (this can be a
     *        compound value, but must be _JSON_ serializable).
     *  (iii) call `dk.trigger(this.widget, 'change', this.widget)` whenever
     *        the filter value changes.
     *        
     */
    _get_filter_data(filters) {
        return Object.entries(this.filters).map(([filtername, filterdef]) => {
            filterdef.name = filtername;
            if (is.isProps(filterdef)) {
                if (filterdef.select_multiple) {  // (a)
                    // console.info("SELECT_MULTIPLE", filterdef);
                    return [filterdef, SelectMultipleFilterDef];
                } else if (filterdef.construct_filter) {  // (b)
                    // console.info("CUSTOMFILTER", filterdef);
                    return [filterdef, CustomFilterDef];
                } else {
                    // console.info("SELECT_ONE", filterdef);
                    return [filterdef, SelectOneFilterDef];
                }
            } else {
                dkconsole.warn("filter def is not a properties object..?");
                // return [filterdef, null]; 
            }
        });
    }

    /**
     * Create widgets, and append to `this.content`, for all pairs in 
     * `this._filter_data`. Set it up so any `change` trigger on the filters 
     * triggers a `filter-change` event on the filter (with `this.values()` 
     * as the only parameter.
     */
    construct() {
        this.filters = {};
        this._filter_data.forEach(([filterprops, filtercls]) => {
            filterprops.datafilter = this;
            const res = filtercls.append_to(this.content, filterprops);
            dk.on(res, 'change', (...args) => {
                if (this.data) this.data.set_filter(this.values());
                this.trigger('filter-change', this.values());
            });
            this.filters[filterprops.name] = res;
        });
    }

    values() {
        const keys = Object.keys(this.filters);
        const vals = keys.map(filter_name => {
            const filter = this.filters[filter_name];
            if (filter.value) {
                // filter has defined a value on itself.. (e.g. `CustomFilterDef`'s
                return filter.value;
            }
            if (filter.input && filter.input.jquery) {
                // found a jQuery object, get its .val()
                return filter.input.val();
            }
            if (filter.input && filter.input.value) {
                // found a dk.form object
                return filter.input.value;
            }
        });
        return zip_object(keys, vals);
    }
}
