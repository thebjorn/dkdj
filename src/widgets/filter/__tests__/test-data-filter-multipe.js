import $ from 'jquery';
import page from "../../../widgetcore/dk-page";
import {DataFilter} from "../data-filter";
import utidy from "../../../browser/dk-html";



test("test-data-filter-multiple", () => {
    document.body.innerHTML = `
        <div id="work"></div>
    `;
    const work = $('#work');
    page.initialize(document);
    let set_filter_vals = null;

    const myfilter = DataFilter.create_inside(work, {
        filters: {
            filter_name: {
                label: 'Filter Label',
                values: [1, 2, 3],
                select_multiple: true
            }
        },
        data: {
            set_filter: function (vals) {
                set_filter_vals = vals;
            }
        }
    });

    console.log(work.html());
    expect(work.html()).toMatchSnapshot();

    work.find(':checkbox:eq(0)').click();
    work.find(':checkbox:eq(1)').click();
    expect(set_filter_vals).toEqual({filter_name: ["1", "2"]});

    // expect(utidy(work.html())).toEqual(utidy(`
    //
    // `));
});
