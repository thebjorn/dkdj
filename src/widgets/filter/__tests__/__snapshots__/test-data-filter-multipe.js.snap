// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-data-filter-multiple 1`] = `
"<section id=\\"data-filter-1\\" class=\\"DataFilter\\"><header id=\\"dk-bx-1\\"><h2 id=\\"dk-bx-2\\" class=\\"title\\">Filter</h2></header><div id=\\"dk-bx-3\\" class=\\"content\\"><div id=\\"select-multiple-filter-def-1\\" class=\\"SelectMultipleFilterDef filterdef\\"><div id=\\"dk-bx-4\\" class=\\"filter-name filterbox\\"><header id=\\"dk-bx-5\\" class=\\"filterheader\\"><h3 id=\\"dk-bx-6\\" class=\\"filtertitle\\">Filter Label</h3></header><div id=\\"dk-bx-7\\" class=\\"filter content filtercontent CheckboxSelectWidget\\"><label id=\\"filter_name-1-label\\" for=\\"filter_name-1\\"><input type=\\"checkbox\\" name=\\"filter_name\\" id=\\"filter_name-1\\" value=\\"1\\">1</label>
<label id=\\"filter_name-2-label\\" for=\\"filter_name-2\\"><input type=\\"checkbox\\" name=\\"filter_name\\" id=\\"filter_name-2\\" value=\\"2\\">2</label>
<label id=\\"filter_name-3-label\\" for=\\"filter_name-3\\"><input type=\\"checkbox\\" name=\\"filter_name\\" id=\\"filter_name-3\\" value=\\"3\\">3</label>
</div></div></div></div></section>"
`;
